src/Workspaces/Core/Portable/FlowAnalysis/SymbolUsageAnalysis/SymbolUsageAnalysis.BasicBlockAnalysisData.cs

```diff
@@ -0,0 +1,178 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.FlowAnalysis.SymbolUsageAnalysis
+{
+    internal static partial class SymbolUsageAnalysis
+    {
+        /// <summary>
+        /// Analysis data for a particular <see cref="BasicBlock"/> for <see cref="ControlFlowGraph"/>
+        /// based dataflow analysis OR for the entire executable code block for high level operation
+        /// tree based analysis.
+        /// </summary>
+        private sealed class BasicBlockAnalysisData : IEquatable<BasicBlockAnalysisData>
+        {
+            private static readonly ObjectPool<BasicBlockAnalysisData> s_pool =
+                new ObjectPool<BasicBlockAnalysisData>(() => new BasicBlockAnalysisData());
+
+            /// <summary>
+            /// Map from each symbol to possible set of reachable write operations that are live at current program point.
+            /// A write is live if there is no intermediate write operation that overwrites it.
+            /// </summary>
+            private readonly Dictionary<ISymbol, PooledHashSet<IOperation>> _reachingWrites;
+
+            private BasicBlockAnalysisData()
+            {
+                _reachingWrites = new Dictionary<ISymbol, PooledHashSet<IOperation>>();
+            }
+
+            public static BasicBlockAnalysisData GetInstance() => s_pool.Allocate();
+
+            public void Free()
+            {
+                FreeAndClearValues();
+                s_pool.Free(this);
+            }
```

Comment:
this seems odd.  why not just be disposable?

Reply:
From my experience and performance analysis of flow analysis in roslyn-analyzers, the primary hotspots for flow analysis are below:
1. Allocations from custom analysis data being tracked for a basic block (these can be quite significant for large flow graphs)
2. Execution time from Merge operation below

I have tried to use pooling here for `BasicBlockAnalysisData` to tackle 1. and tried to keep Merge as minimal as possible for 2.

---
In reply to: [231669592](https://github.com/dotnet/roslyn/pull/31025#discussion_r231669592) [](ancestors = 231669592)